In [54]:
import pandas as pd

In [55]:
data=pd.read_csv('data00_100000_clean.csv')

In [56]:
data=data.drop(['AQS_Code', 'year'], axis=1)

In [57]:
data=data.drop(['date'],axis=1)

In [58]:
data

Latitude  Longitude      epoch        no       no2       nox  \
0      29.767997 -95.220582  949363200   1.84932  25.62468  28.03901   
1      29.767997 -95.220582  949363500   0.94965  23.56869  24.83741   
2      29.767997 -95.220582  949363800   0.84969  26.22643  27.53876   
3      29.767997 -95.220582  949364100   0.84969  29.28535  30.64030   
4      29.767997 -95.220582  949364400   1.14958  29.28535  30.94046   
5      29.767997 -95.220582  949364700   1.74936  32.14367  34.64230   
6      29.767997 -95.220582  949365000   1.74936  32.29412  34.74235   
7      29.767997 -95.220582  949365300   1.74936  33.04631  35.34265   
8      29.767997 -95.220582  949365600   2.34914  33.14660  35.74286   
9      29.767997 -95.220582  949365900   1.74936  29.38564  31.64081   
10     29.767997 -95.220582  949366200   1.04961  27.93140  29.63981   
11     29.767997 -95.220582  949366500   1.84932  32.19382  34.64230   
12     29.767997 -95.220582  949366800   1.44947  29.28535  31.34065   
13     29.767997 -95.220582  949367100   1.84932  33.34719  35.44270   
14     29.767997 -95.220582  949367400   1.64939  35.75420  37.64380   
15     29.767997 -95.220582  949367700   1.24954  32.39441  34.24210   
16     29.767997 -95.220582  949368000   3.14884  41.42071  44.54725   
17     29.767997 -95.220582  949368300   4.14848  44.52977  48.84940   
18     29.767997 -95.220582  949368600   3.74862  40.26735  44.04700   
19     29.767997 -95.220582  949368900   1.24954  32.24397  34.04200   
20     29.767997 -95.220582  949369200   1.54943  30.63929  32.54125   
21     29.767997 -95.220582  949369500   0.84969  26.32672  27.73886   
22     29.767997 -95.220582  949369800   0.84969  25.52439  27.13856   
23     29.767997 -95.220582  949370100   0.84969  22.66606  23.73686   
24     29.767997 -95.220582  949370400   1.74936  21.71328  24.13706   
25     29.767997 -95.220582  949370700   1.04961  24.62176  25.83791   
26     29.767997 -95.220582  949371000   0.84969  24.67190  25.63781   
27     29.767997 -95.220582  949371300   0.84969  22.66606  23.93696   
28     29.767997 -95.220582  949371600   0.04998  21.06138  21.83591   
29     29.767997 -95.220582  949371900   0.84969  21.61299  22.93646   
...          ...        ...        ...       ...       ...       ...   
84388  29.767997 -95.220582  976675500   9.44464  16.68757  25.96683   
84389  29.767997 -95.220582  976675800  14.85564  20.02509  34.58978   
84390  29.767997 -95.220582  976676100  10.03493  18.99438  28.61250   
84391  29.767997 -95.220582  976676400   7.18186  17.03114  24.00707   
84392  29.767997 -95.220582  976676700   9.24788  17.22746  26.26079   
84393  29.767997 -95.220582  976677000   6.00128  15.90228  21.45938   
84394  29.767997 -95.220582  976677300   6.78834  15.26422  21.65536   
84395  29.767997 -95.220582  976677600   6.88672  15.90228  22.43926   
84396  29.767997 -95.220582  976677900   5.60776  17.47287  22.63523   
84397  29.767997 -95.220582  976678200   8.75597  20.71222  29.29842   
84398  29.767997 -95.220582  976678500  15.93783  24.78596  40.17509   
84399  29.767997 -95.220582  976678800  17.41356  24.44239  41.25296   
84400  29.767997 -95.220582  976679100   8.95273  20.07417  28.80848   
84401  29.767997 -95.220582  976679400   3.73850  15.06790  18.32376   
84402  29.767997 -95.220582  976679700   8.85435  17.52195  26.06482   
84403  29.767997 -95.220582  976680000   8.46083  17.76736  25.86884   
84404  29.767997 -95.220582  976680300   4.91908  16.93298  21.55737   
84405  29.767997 -95.220582  976680600   7.08348  18.25817  24.88896   
84406  29.767997 -95.220582  976680900   7.08348  20.02509  26.75073   
84407  29.767997 -95.220582  976681200   6.00128  19.77968  25.47689   
84408  29.767997 -95.220582  976681500   8.85435  21.98833  30.67026   
84409  29.767997 -95.220582  976681800  10.03493  24.19698  33.90386   
84410  29.767997 -95.220582  976682100   8.46083  23.85342  32.04208   
84411  29.767997 -95.220582  976682400  11.21551

In [59]:
def time_to_num(time):
    splits=time.split(':')
    num=int(splits[0])*60+int(splits[1])
    return(int(num))

In [60]:
time_to_num('06:15:00')

375

In [61]:
data['time']=data['time'].apply(time_to_num)

In [62]:
data['time']

0          0
1          5
2         10
3         15
4         20
5         25
6         30
7         35
8         40
9         45
10        50
11        55
12        60
13        65
14        70
15        75
16        80
17        85
18        90
19        95
20       100
21       105
22       110
23       115
24       120
25       125
26       130
27       135
28       140
29       145
        ... 
84388    165
84389    170
84390    175
84391    180
84392    185
84393    190
84394    195
84395    200
84396    205
84397    210
84398    215
84399    220
84400    225
84401    230
84402    235
84403    240
84404    245
84405    250
84406    255
84407    260
84408    265
84409    270
84410    275
84411    280
84412    285
84413    290
84414    295
84415    300
84416    305
84417    310
Name: time, Length: 84418, dtype: int64

In [66]:
data=data.drop(['time'], axis=1)

In [79]:
data_strs=data
data_strs['day']=data_strs['day'].apply(lambda x: str(x))
data_strs['month']=data_strs['month'].apply(lambda x: str(x))
data_strs['hour']=data_strs['hour'].apply(lambda x: str(x))

In [80]:
dummies=pd.get_dummies(data_strs)

In [82]:
dummies.columns

Index(['Latitude', 'Longitude', 'epoch', 'no', 'no2', 'nox', 'o3', 'winddir',
       'windspd', 'wind_x_dir', 'wind_y_dir', 'day_1', 'day_10', 'day_11',
       'day_12', 'day_13', 'day_14', 'day_15', 'day_16', 'day_17', 'day_18',
       'day_19', 'day_2', 'day_20', 'day_21', 'day_22', 'day_23', 'day_24',
       'day_25', 'day_26', 'day_27', 'day_28', 'day_29', 'day_3', 'day_30',
       'day_31', 'day_4', 'day_5', 'day_6', 'day_7', 'day_8', 'day_9',
       'month_10', 'month_11', 'month_12', 'month_2', 'month_3', 'month_4',
       'month_5', 'month_6', 'month_7', 'month_8', 'month_9', 'hour_0',
       'hour_1', 'hour_10', 'hour_11', 'hour_12', 'hour_13', 'hour_14',
       'hour_15', 'hour_16', 'hour_17', 'hour_18', 'hour_19', 'hour_2',
       'hour_20', 'hour_21', 'hour_22', 'hour_23', 'hour_3', 'hour_4',
       'hour_5', 'hour_6', 'hour_7', 'hour_8', 'hour_9'],
      dtype='object')

In [69]:
data

Latitude  Longitude      epoch        no       no2       nox  \
0      29.767997 -95.220582  949363200   1.84932  25.62468  28.03901   
1      29.767997 -95.220582  949363500   0.94965  23.56869  24.83741   
2      29.767997 -95.220582  949363800   0.84969  26.22643  27.53876   
3      29.767997 -95.220582  949364100   0.84969  29.28535  30.64030   
4      29.767997 -95.220582  949364400   1.14958  29.28535  30.94046   
5      29.767997 -95.220582  949364700   1.74936  32.14367  34.64230   
6      29.767997 -95.220582  949365000   1.74936  32.29412  34.74235   
7      29.767997 -95.220582  949365300   1.74936  33.04631  35.34265   
8      29.767997 -95.220582  949365600   2.34914  33.14660  35.74286   
9      29.767997 -95.220582  949365900   1.74936  29.38564  31.64081   
10     29.767997 -95.220582  949366200   1.04961  27.93140  29.63981   
11     29.767997 -95.220582  949366500   1.84932  32.19382  34.64230   
12     29.767997 -95.220582  949366800   1.44947  29.28535  31.34065   
13     29.767997 -95.220582  949367100   1.84932  33.34719  35.44270   
14     29.767997 -95.220582  949367400   1.64939  35.75420  37.64380   
15     29.767997 -95.220582  949367700   1.24954  32.39441  34.24210   
16     29.767997 -95.220582  949368000   3.14884  41.42071  44.54725   
17     29.767997 -95.220582  949368300   4.14848  44.52977  48.84940   
18     29.767997 -95.220582  949368600   3.74862  40.26735  44.04700   
19     29.767997 -95.220582  949368900   1.24954  32.24397  34.04200   
20     29.767997 -95.220582  949369200   1.54943  30.63929  32.54125   
21     29.767997 -95.220582  949369500   0.84969  26.32672  27.73886   
22     29.767997 -95.220582  949369800   0.84969  25.52439  27.13856   
23     29.767997 -95.220582  949370100   0.84969  22.66606  23.73686   
24     29.767997 -95.220582  949370400   1.74936  21.71328  24.13706   
25     29.767997 -95.220582  949370700   1.04961  24.62176  25.83791   
26     29.767997 -95.220582  949371000   0.84969  24.67190  25.63781   
27     29.767997 -95.220582  949371300   0.84969  22.66606  23.93696   
28     29.767997 -95.220582  949371600   0.04998  21.06138  21.83591   
29     29.767997 -95.220582  949371900   0.84969  21.61299  22.93646   
...          ...        ...        ...       ...       ...       ...   
84388  29.767997 -95.220582  976675500   9.44464  16.68757  25.96683   
84389  29.767997 -95.220582  976675800  14.85564  20.02509  34.58978   
84390  29.767997 -95.220582  976676100  10.03493  18.99438  28.61250   
84391  29.767997 -95.220582  976676400   7.18186  17.03114  24.00707   
84392  29.767997 -95.220582  976676700   9.24788  17.22746  26.26079   
84393  29.767997 -95.220582  976677000   6.00128  15.90228  21.45938   
84394  29.767997 -95.220582  976677300   6.78834  15.26422  21.65536   
84395  29.767997 -95.220582  976677600   6.88672  15.90228  22.43926   
84396  29.767997 -95.220582  976677900   5.60776  17.47287  22.63523   
84397  29.767997 -95.220582  976678200   8.75597  20.71222  29.29842   
84398  29.767997 -95.220582  976678500  15.93783  24.78596  40.17509   
84399  29.767997 -95.220582  976678800  17.41356  24.44239  41.25296   
84400  29.767997 -95.220582  976679100   8.95273  20.07417  28.80848   
84401  29.767997 -95.220582  976679400   3.73850  15.06790  18.32376   
84402  29.767997 -95.220582  976679700   8.85435  17.52195  26.06482   
84403  29.767997 -95.220582  976680000   8.46083  17.76736  25.86884   
84404  29.767997 -95.220582  976680300   4.91908  16.93298  21.55737   
84405  29.767997 -95.220582  976680600   7.08348  18.25817  24.88896   
84406  29.767997 -95.220582  976680900   7.08348  20.02509  26.75073   
84407  29.767997 -95.220582  976681200   6.00128  19.77968  25.47689   
84408  29.767997 -95.220582  976681500   8.85435  21.98833  30.67026   
84409  29.767997 -95.220582  976681800  10.03493  24.19698  33.90386   
84410  29.767997 -95.220582  976682100   8.46083  23.85342  32.04208   
84411  29.767997 -95.220582  976682400  11.21551

In [72]:
from keras.utils import np_utils

In [77]:
day_cat=np_utils.to_categorical(data['day'])b
month_cat=np_utils.to_categorical(data['month'])
hour_cat=np_utils.to_categorical(data['hour'])

In [83]:
# Use numpy to convert to arrays
import numpy as np
# Labels are the values we want to predict
labels = np.array(dummies['o3'])
# Remove the labels from the features
# axis 1 refers to the columns
features= dummies.drop('o3', axis = 1)
# Saving feature names for later use
feature_list = list(features.columns)
# Convert to numpy array
features = np.array(features)

In [84]:
# Using Skicit-learn to split data into training and testing sets
from sklearn.model_selection import train_test_split
# Split the data into training and testing sets
train_features, test_features, train_labels, test_labels = train_test_split(features, labels, test_size = 0.25, random_state = 1)

<h3><i>Establish baseline?

<h2>Model Implementation Time</h2>

In [86]:
# Import the model we are using
from sklearn.ensemble import RandomForestRegressor
# Instantiate model with 1000 decision trees
rf = RandomForestRegressor(n_estimators = 1000, random_state = 42)
#use verbose=1
# Train the model on training data
rf.fit(train_features, train_labels);

C:\Users\Nicholas\AppData\Local\Programs\Python\Python36\lib\site-packages\sklearn\ensemble\weight_boosting.py:29: DeprecationWarning: numpy.core.umath_tests is an internal NumPy module and should not be imported. It will be removed in a future NumPy release.
  from numpy.core.umath_tests import inner1d


In [89]:
import pickle

pickle.dump(rf, open('rf_1.pkl', 'wb'))

MemoryError: 

In [91]:
from sklearn.externals import joblib

# Save to file in the current working directory
joblib_file = "rf_1.pkl"  
joblib.dump(rf, joblib_file)

['rf_1.pkl']

In [94]:
rf.score(test_features, test_labels)

0.9516668132093452

In [96]:
from sklearn import model_selection
kfold = model_selection.KFold(n_splits=10, random_state=0)
scoring = 'neg_mean_absolute_error'
results = model_selection.cross_val_score(rf, test_features, test_labels, cv=kfold, scoring=scoring)
results

array([-4.19007583, -4.15650348, -4.06492095, -3.97509023, -4.09298393,
       -4.21514153, -4.32852868, -4.11073779, -4.144749  , -3.97688133])

In [99]:
results.mean()

-4.1255612746321315

<i>R-squared score on 1000 trees: 0.9516668132093452
    
Mean Absolute Error scores (cross-val): array([-4.19007583, -4.15650348, -4.06492095, -3.97509023, -4.09298393,
       -4.21514153, -4.32852868, -4.11073779, -4.144749  , -3.97688133])
       mean MAE: -4.1255612746321315

<h3>Notes</h3>

Try with lower tree count, changing the maximal depth

In [101]:
rf.n_features_

76

In [104]:
feat_importances=rf.feature_importances_

In [107]:
feature_importances = pd.DataFrame(feat_importances,
                                   index = dummies.drop('o3', axis=1).columns,
                                    columns=['importance']).sort_values('importance', ascending=False)

In [108]:
feature_importances

importance
no            0.320272
no2           0.089894
epoch         0.085368
wind_y_dir    0.053057
hour_20       0.034007
hour_19       0.033746
hour_18       0.033731
windspd       0.029440
hour_21       0.029245
hour_17       0.026219
wind_x_dir    0.023916
hour_22       0.023553
hour_16       0.019468
nox           0.018600
hour_23       0.017232
month_7       0.016995
month_8       0.016222
winddir       0.015738
hour_0        0.010852
hour_15       0.010214
month_4       0.006709
hour_1        0.004727
hour_14       0.004620
day_31        0.003619
day_7         0.003249
month_9       0.003236
month_6       0.003064
day_19        0.003028
day_5         0.002864
day_16        0.002784
...                ...
day_22        0.001678
hour_2        0.001280
hour_13       0.001276
month_3       0.001215
day_24        0.001209
day_29        0.001170
day_27        0.001143
day_10        0.001140
day_30        0.001031
day_4         0.001025
month_10      0.000994
month_11      0.000939
day_6         0.000762
day_28        0.000750
day_3         0.000702
day_11        0.000691
hour_3        0.000573
month_2       0.000349
hour_12       0.000273
hour_11       0.000263
hour_10       0.000234
hour_4        0.000218
hour_8        0.000185
hour_9        0.000175
hour_5        0.000171
hour_6        0.000150
hour_7        0.000126
month_12      0.000077
Longitude     0.000000
Latitude      0.000000

[76 rows x 1 columns]

In [113]:
from sklearn.grid_search import GridSearchCV

rf_reg=RandomForestRegressor()
parameters= {'n_estimators':[100], 'max_features':[2,4,8,16,32,64,76], 'min_samples_leaf': [1,10,50,100,200,500], 'min_samples_split': [2,4,8,16]}
rf_grid = GridSearchCV(rf_reg, parameters, cv = 4)
rf_grid.fit(train_features, train_labels)

GridSearchCV(cv=4, error_score='raise',
       estimator=RandomForestRegressor(bootstrap=True, criterion='mse', max_depth=None,
           max_features='auto', max_leaf_nodes=None,
           min_impurity_decrease=0.0, min_impurity_split=None,
           min_samples_leaf=1, min_samples_split=2,
           min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=1,
           oob_score=False, random_state=None, verbose=0, warm_start=False),
       fit_params={}, iid=True, n_jobs=1,
       param_grid={'n_estimators': [100], 'max_features': [2, 4, 8, 16, 32, 64, 76], 'min_samples_leaf': [1, 10, 50, 100, 200, 500], 'min_samples_split': [2, 4, 8, 16]},
       pre_dispatch='2*n_jobs', refit=True, scoring=None, verbose=0)

In [116]:
rf_100=rf_grid.best_estimator_

In [119]:
rf_100.score(test_features, test_labels)

0.9517633884882108

In [120]:
# Save to file in the current working directory
joblib_file = "rf_100.pkl"  
joblib.dump(rf_100, joblib_file)

['rf_100.pkl']

In [121]:
rf_10=RandomForestRegressor(n_estimators=10)
rf_10.fit(train_features, train_labels)

RandomForestRegressor(bootstrap=True, criterion='mse', max_depth=None,
           max_features='auto', max_leaf_nodes=None,
           min_impurity_decrease=0.0, min_impurity_split=None,
           min_samples_leaf=1, min_samples_split=2,
           min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=1,
           oob_score=False, random_state=None, verbose=0, warm_start=False)

In [124]:
rf_10.score(test_features, test_labels)

0.9426992188147277

In [125]:
joblib_file = "rf_10.pkl"  
joblib.dump(rf_10, joblib_file)

['rf_10.pkl']

In [114]:
rf_grid.best_params_

{'max_features': 64,
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'n_estimators': 100}

In [126]:
# Use the forest's predict method on the test data
predictions = rf_10.predict(test_features)
# Calculate the absolute errors
errors = abs(predictions - test_labels)
# Print out the mean absolute error (mae)
print('Mean Absolute Error:', round(np.mean(errors), 2))

Mean Absolute Error: 3.07


# Load from file
joblib_model = joblib.load(joblib_file)

# Calculate the accuracy and predictions
score = joblib_model.score(Xtest, Ytest)  
print("Test score: {0:.2f} %".format(100 * score))  
Ypredict = pickle_model.predict(Xtest)  